### DISCUSSION TITLE
Zooz ZSE19 Siren/Chime Sound Switch Question

### DISCUSSION THREAD

**Discussion by GrizzlyAK** [zwave-js/zwave-js#3161] [Q&A] [answered]
Env: Home Assistant ZWave JS Integration for Websocket connection to zwavejs2mqtt Add On for Control Panel only.

I have the subject device, which presents in HA as a binary switch for the Siren (works perfectly) and a NOTHING for the Chime. Both Siren and Sound Switch are implemented in the device and are independent. The SS section in the zwavejs2mqtt Control Panel provides the following four parameters: Default Volume, Default Tone ID, Play Tone (a list of sounds in the device at 5 seconds only), and Volume (Volume only provides two options: Default and Custom - but you can't enter a number - this could be a CP issue.)

Using the CP, I can select a sound in the Play Tone parameter and sound that Chime, at the volume I set in the Default Volume parameter. 

I have two issues: 

(1) After setting a Default Volume (which it does play at) and a Default Tone ID, whenever I have to refresh the page (or after some time, I don't know which), these values reset to 100% and 1, respectively. Should they not be static and reflect what is stored in the device? It appears the device reverts as well. What could be going on here?
(2) I'm not sure how exactly to use the Sound Switch in HA in an automation, although the following works to play a tone, I've been unsuccessful in using similar code to set the Default Volume or any other parameters. I also thought that SS allowed for setting Tone and Volume during the call. Perhaps this is a HA shortcoming, idk.

```  
     - service: zwave_js.set_value
        data:
          command_class: '121'
          endpoint: '0'
          property: toneId
          value: 34
        target:
          entity_id: sensor.zooz_zse19_siren_node_status
```

I realize some of this may not be the result of the library, but I'm just trying to get my head around what's going on and how best to make this work. Any advice would be appreciated. Thanks.

**Comment by AlCalzone** [comment]
> Volume (Volume only provides two options: Default and Custom - but you can't enter a number - this could be a CP issue

This is an issue in zwavejs2mqtt. `Sound Switch CC` defines this metadata for the volume:
```ts
@ccValueMetadata({
	...ValueMetadata.UInt8,
	min: 0,
	max: 100,
	unit: "%",
	label: "Volume",
	states: {
		0: "default",
	},
})
public volume?: number;
```
and zwavejs2mqtt seems to think that only the `0` is allowed instead of the full range. @robertsLando 

> After setting a Default Volume (which it does play at) and a Default Tone ID, whenever I have to refresh the page (or after some time, I don't know which), these values reset to 100% and 1, respectively. Should they not be static and reflect what is stored in the device? It appears the device reverts as well. What could be going on here?

Not sure. Will need to see a driver log (level "debug") of when that happens. Maybe the device sends an update for these.

> I've been unsuccessful in using similar code to set the Default Volume or any other parameters.

You should be able to using the properties `defaultToneId` and `defaultVolume`. Setting the `volume` only doesn't make a lot of sense because this won't communicate with the device. You can pass the `volume` as an option when setting the `toneId` though, but I don't know how the syntax for that is in HA.

**Reply by GrizzlyAK** [reply]
re: "Volume",  at the top of the MQTT CP, after trying to enter a value for Volume in Sound Swtich, I see the following: ` Driver: Sound Switch: "volume" is not a supported property`, which I assume is coming from the mqtt side. I tried setting the volume when calling the Set Value in the SS CC in HA Dev Tools, but it had no effect. The defaultVolume value DOES work, however, once set. cc @robertsLando

**Reply by AlCalzone** [reply]
That's coming from the driver. Thinking about this error, I'm not very happy with how this works.

On the device, `volume` can only be set together with `toneId`, which is why it isn't "supported". Allowing it however would break a convention of these `setValue` calls because this would be the only one that doesn't communicate with the device. It would be the only way though to set it without a specialized UI for sound switches.

**Reply by GrizzlyAK** [reply]
That makes sense.

**Comment by robertsLando** [comment]
> and zwavejs2mqtt seems to think that only the 0 is allowed instead of the full range. @robertsLando

If he writes the value in the input and press the send button it will send the value, it just shows a list of the defined states

**Comment by kpine** [comment, marked as answer]
The HA `zwave_js.set_value` service call is basically a pass-through to the node-zwave-js `setValue` API. If you want to set the volume when you play a tone, you need to pass in the volume as an option, just as you would with the node-zwave-js API. I believe this should set the volume to 50% (notice the added `options` map):

```yaml
     - service: zwave_js.set_value
        data:
          command_class: '121'
          endpoint: '0'
          property: toneId
          value: 34
          options:
            volume: 50
        target:
          entity_id: sensor.zooz_zse19_siren_node_status
```

With the 2021.8 release, HA supports Sound Switch CC-based [sirens](https://www.home-assistant.io/integrations/siren/) natively, so there's no need to know about these specific API details. You can play tones and set volume using platform services. The ability to set default volume and default tone is still a work in progress, so for now you'll need to still use `set_value` for those.

**Reply by GrizzlyAK** [reply]
Thank you. It appears that the Volume option for this call is not yet implemented in zwavejs2mqtt (according to a message that appears in the CP when trying to use it there), and my tests of the yaml above in Dev Tools Services yields no changes in volume played. Only the defaultVolume value has any affect atm.

**Reply by kpine** [reply]
Perhaps you should post some logs then. There's no indication that there's anything wrong from the HA point of view.

**Reply by kpine** [reply]
Volume is only an option for V2 of the Sound Switch CC. According to the product DB, the Zooz does not specify a version, which usually means V1. That would mean your siren does not support setting the volume. Can you confirm the version? Maybe node-zwave-js is creating a volume value even for V1 CC?

**Reply by GrizzlyAK** [reply]
True, Zooz manual just lists COMMAND_CLASS_SOUND_SWITCH. So... V1. The Ecolink Siren+Chime ALSO lists V1 for SS CC. Perhaps THAT is where that message is coming from in the zw2m CP and not the Add-on itself (which is what I initially thought).

I've been going back and forth with Zooz about their SS functionality and the chime randomly ringing multiple times (1-10), upon trigger (occasionally). They blamed it on OZW1.4 and me using the Set Config Param function to sound the chime as the cause, but it is happening using the new SS CC under zwave-js as well, so... 

And, sadly, I got the following email from them today, indicating the Siren is being discontinued.

> The sound switch command class is implemented in a standard way in the device and is documented starting page 562 in the Z-Wave command class guide here:
https://www.silabs.com/documents/login/miscellaneous/SDS13781-Z-Wave-Application-Command-Class-Specification.pdf
It does look like a problem on the siren side but because we haven't been able to replicate it on any other platform and from my understanding, the multiple sounds are quite inconsistent, we probably won't be able to address it. I was also notified that the product will be discontinued in the near future so a firmware or documentation update is unlikely.

**Reply by kpine** [reply]
> Perhaps THAT is where that message is coming from in the zw2m CP and not the Add-on itself (which is what I initially thought).

No, the message comes from the driver as stated in this [comment](https://github.com/zwave-js/node-zwave-js/discussions/3161#discussioncomment-1136901).

The zwavejs2mqtt or HA debug dumps should show you the CC version. If the version is V1, then you're out of luck.

**Reply by GrizzlyAK** [reply]
> No, the message comes from the driver as stated in this comment.

Yes. That's what I was saying. The CP is just displaying the Driver message, which generated it. I originally thought that it was zw2m which was generating the message (and didn't implement it, blocking HA from sending it to the driver), until the AlCalzone's comment cleared it up.

**Reply by kpine** [reply]
> zw2m which was generating the message (and didn't implement it, blocking HA from sending it to the driver)

Just to clarify, when using the integration HA has no interaction with zwavejs2mqtt at all. zwavejs2mqtt is hosting the websocket zwave-js server. All interaction using the integration goes through that websocket server and is not gated by zwavejs2mqtt in any way.

**Reply by GrizzlyAK** [reply]
Cool! Thanks for that clarification. Still trying to wrap my head around all the pieces. Thanks for taking the time to educate me. :->

So, when I change a parameter using the zwavejs2mqtt CP, it communicates with the driver, which is then picked up via WS by HA? And vice/versa?
